### ISSUE TITLE
Get TenantInfo in OnConfiguring method of ApplicationDbContext

### ISSUE THREAD

**Issue by endrihacker** [Finbuckle/Finbuckle.MultiTenant#331]
I am using separate databases for each tenant and I am following the docs [here](https://www.finbuckle.com/MultiTenant/Docs/EFCore#separate-databases). 
I can't make it work to have TenantInfo ConnectionString on OnConfiguring method of ApplicationDbContext. Dependency injection doesn't seem to return a tenant or default connection string if tenant is missing. 

How to ge

Also the sample **IdentityDataIsolationSample** doesn't seem to work. It throws an error that TenantInfo is null at this line `optionsBuilder.UseSqlite(TenantInfo.ConnectionString);` of `ApplicationDbContext`.

My code looks like this:

    public class ApplicationDbContext : IdentityDbContext<ApplicationUser, ApplicationRole, string>
    {
        private ITenantInfo TenantInfo { get; set; }

        ....

        public ApplicationDbContext(DbContextOptions options, ITenantInfo tenantInfo) : base(options)
        {
            TenantInfo = tenantInfo;
        }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer(TenantInfo.ConnectionString, b => b.MigrationsAssembly("MyApp"));
            base.OnConfiguring(optionsBuilder);
        }
    }

How can get value of `TenantInfo`?

**Comment by AndrewTriesToCode** [comment]
Hi @endrihacker,

I just tried the sample app and didn't run into the error you mentioned. Can you post your app's `Startup` methods so I can see how you are set up?

**Comment by endrihacker** [comment]
Hi @AndrewTriesToCode,

thanks for your reply!

I just realized that I was at the branch `release-6.0.0-dev` that's where it doesn't work. 

I'm starting from [QuickApp](https://github.com/emonney/QuickApp) template and most of the code is from the last [commit](https://github.com/emonney/QuickApp/commit/0a082a3cdb9e9d1a75bc53f470298b089b92924c).

The changes i made are in the snippet code below

    public void ConfigureServices(IServiceCollection services)
    {
        // Moved up these two methods
        // Add cors
        services.AddCors();

        services.AddControllersWithViews();

        // Configure Multitenancy
        services.AddMultiTenant<TenantInfo>()
                .WithConfigurationStore()
                .WithRouteStrategy();

        services.AddDbContext<ApplicationDbContext>();

        // add identity
        services.AddIdentity<ApplicationUser, ApplicationRole>()
            .AddEntityFrameworkStores<ApplicationDbContext>()
            .AddDefaultTokenProviders();

        // The rest is as in the QuickApp except for what I moved at the beginning                
    }


    // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
    public void Configure(IApplicationBuilder app, IWebHostEnvironment env, ILoggerFactory loggerFactory)
    {
        //...

        app.UseRouting();

        // Use Multitenancy
        app.UseMultiTenant();

        app.UseCors(builder => builder
            .AllowAnyOrigin()
            .AllowAnyHeader()
            .AllowAnyMethod());

        // ...
        app.UseEndpoints(endpoints =>
        {
            endpoints.MapControllerRoute(
                name: "tenant",
                pattern: "{__tenant__=}/{controller}/{action=Index}/{id?}");
            endpoints.MapControllerRoute(
                name: "default",
                pattern: "{controller}/{action=Index}/{id?}");
        });

        // ...
    }


But today I have things more clear. I just get TenantInfo only when a tenant is found from the strategy I use. I thought that I can get the default `ConnectionString` even when no tenant is determined. I skipped the fact that it is null when no tenant is found. This is what I did wrong, in case someone else read this thread.  So what I changed is:

At `Startup`

From: 

    services.AddDbContext<ApplicationDbContext>();

to

    services.AddDbContext<ApplicationDbContext>(options =>
                options.UseSqlServer(Configuration["ConnectionStrings:DefaultConnection"], b => b.MigrationsAssembly("QuickApp")));

as it was in the beginning template.

And changed the connection string only if a tenant is determined

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
    {
        if (TenantInfo != null)
        {
            optionsBuilder.UseSqlServer(TenantInfo.ConnectionString, b => b.MigrationsAssembly("QuickApp"));
        }
        base.OnConfiguring(optionsBuilder);
    }


This solved what I needed to do. Does it looks right?

**Comment by AndrewTriesToCode** [comment]
@endrihacker That looks like a good approach. Another thing you can do is define a "fallback" tenant by using the `WithStaticStrategy` method *after* your other strategies. Then its connection string would be used if no other tenant was selected. This might be better if there were other defaults you wanted for no real tenant, but for connection string your approach works well.

**Comment by endrihacker** [comment]
@AndrewTriesToCode I tied `WithStaticStrategy` but since I have decorated the controllers with `[Route("{__tenant__}/api/[controller]")]` attribute but this way `{__tenant__}` doesn't seems to be optional. So I need to provide a tenant always.

However I tried also to add two route attributes like this

    [Route("/api/[controller]")]
    [Route("{__tenant__}/api/[controller]")]

and this way I can have an optional tenant but because they are two different routes, they appear twice in swagger for the same methods.

I know that parameters can be optional by adding `?` at the end like this `{__tenant__?}` but in this case it doesn't work.

I'd appreciate if you know a better way to solve this. 
However this is off topic, and you can close the issue because I found what I asked for.
Thank you!

**Comment by AndrewTriesToCode** [comment]
It seems like you have a pretty good work-around-- another approach might be to set a default value for the route parameter which would effectively be the "default" tenant.

I don't use route attributes that much, but I have used your approach before without any problems.

**Comment by endrihacker** [comment]
Hello @AndrewTriesToCode 

Now I am more familiar with Finbuckle.
I'm building a rest API and I want to authenticate users and allow them to see only their data.

What I'm doing so far is to use the `/connect/token` endpoint to authenticate users and use `WithRouteStrategy()` to differentiate tenants later. But this approach seems to have security issues be cause malicious users can change tenant identifier to get other user data.

So reading more issues here I thought that I need to use `WithClaimStrategy()`. That way the users can't modify the tenant.

I have the following setup:

Multitenant Configuration:

	services.AddMultiTenant<ExtendedTenantInfo>()
	        //.WithConfigurationStore()
	        .WithEFCoreStore<MultiTenantStoreDbContext, ExtendedTenantInfo>()
	        .WithClaimStrategy()
	        .WithRouteStrategy()
	        .WithStaticStrategy("finbuckle");

JWT payload data:

	{
	  "nbf": 1603705533,
	  "exp": 1603709133,
	  "iss": "https://localhost:44350",
	  "aud": "quickapp_api",
	  "client_id": "quickapp_spa",
	  "sub": "7dea9638-b38f-424f-9e2e-d9f61d0f99bc",
	  "auth_time": 1603705530,
	  "idp": "local",
	  "role": "administrator",
	  "permission": [
	    "users.view",
	    "users.manage",
	    "roles.view",
	    "roles.manage",
	    "roles.assign"
	  ],
	  "name": "admin",
	  "email": "admin@example.com",
	  "phone_number": "+1 (123) 000-0000",
	  "fullname": "Inbuilt Administrator",
	  "__tenant__": "initech",
	  "scope": [
	    "email",
	    "openid",
	    "phone",
	    "profile",
	    "roles",
	    "quickapp_api",
	    "offline_access"
	  ],
	  "amr": [
	    "pwd"
	  ]
	}

Does the `__tenant__` in the payload data is set correctly? I'm not able to determine the tenant from Claims Strategy. The only tenants I get are from the static strategy, or route strategy if I provide one. Do I need to make any extra change somewhere?

Since I'm using IdentityServer4 I saw the following paragraph from the [docs](https://identityserver4.readthedocs.io/en/release/reference/interactionservice.html):

> Tenant
The tenant requested. This is provided via the “tenant:” prefix to the acr_values parameter on the authorize request.

Is there any relation between `acr_values` and `__tenant__` that you use in this library?

P.S. Static Strategy is for authenticating users against a default database, and tenant is set explicitly in the code (I set it to another tenant different than the default). I will see any option to get it while using the `/connect/token` endpoint. Perhaps `acr_values` can help me, but I need to try it.

**Comment by AndrewTriesToCode** [comment]
hi @endrihacker 

I'll have to test out the claims strategy more with JWT claims -- I mainly use it for cookie-based claims. Do you have a sample project on github that shows what you describe?

The `acr_values` on with IdentityServer is a good way to pass the tenant for an authentication challenge and can ultimately be used by IdentityServer to help set the `__tenant__` claim in the token.

Back in your rest api for each normal request you won't be using `acr_values` except only when you set it to redirect a challenge response. Instead you should be able to use the `__tenant__` claim, but like I said I will test this further.

**Comment by endrihacker** [comment]
@AndrewTriesToCode I have a sample and I will try to put it on github shortly.
Thanks for your support!

**Comment by endrihacker** [comment]
Hi @AndrewTriesToCode !

I just pushed the code at [this](https://github.com/endrihacker/QuickAppMultitenant) repository. It uses SQL Server Express database actually. If you want me to make some changes to make it more easy to run the project let me know.

The current project after running creates the database `QuickAppMDb`. To create other databases, `QuickAppM2d`, `QuickAppM3d` or `QuickAppM1` just for testing purposes I rerun the project by changing database name in `DefaultConnection` in `appsettings.json` file. (I run 3 times to create 3 more databases).

Default username and password are: `admin`, `tempP@ss1234`.
It's good to make some changes at `DAL\DatabaseInitializer.cs` before creating new databases to make them different from each other... like changing the `cust_1` Name each time you create a tenant database.

Here I changed to `WithConfigurationStore` to omit the tenant's database.

Thanks for your time :)

**Comment by endrihacker** [comment]
Hello @AndrewTriesToCode!

Is there any plan to implement claim strategy for JWT claims soon? 
I'm trying to resolve tenant from JWT claims but not any luck so far. Can you suggest me some tips how could I solve it?

I was thinking to get the tenant from any of those implemented strategies like `WithRouteStrategy` or `WithDelegateStrategy` (claim from headers) and later compare the tenant with the tenant claim stored in JWT token but I don't know where to do the comparison. I thought of doing it by overriding `AuthorizeAttribute`  but I see it can't be done in .NET core.

I know this question is outside the scope of this library... but I'm lost and I don't know how to solve this issue!

**Comment by AndrewTriesToCode** [comment]
hi @endrihacker 
The claim strategy in theory should work with JWT claims currently. The only detail is that authentication scheme you are using fot JWT authentication needs to be the default authenticate scheme. If you use the OpenId Connect scheme it does this  for you internally I believe. Also you need to have a claim with a type of `__tenant__` (by default this is used but it can be changed) whose value will be the tenant identifier.

**Comment by AndrewTriesToCode** [comment]
OK I took a closer look at your repo and have some more specific feedback:

In Identity for your user add a `__tenant__` claim using `UserManager.AddClaimAsync`. See my the `Program.cs` file in my Claims Strategy sample to see how I add the claim. I think you'll want to do that in your `CreateUser` method in your database initializer and whereever else you create a new user.

Then when Identity creates the user it will add the new claim. Then I think IdentityServer will copy it to the JWT for you--I'm not 100% sure about that though.

Then in the app if a JWT token for the tenant is present the claim strategy should be able to pick it up.

**Comment by endrihacker** [comment]
Hi @AndrewTriesToCode 

> The only detail is that authentication scheme you are using fot JWT authentication needs to be the default authenticate scheme

I didn't understand well what you mean with that sentence. I don't know if it was still relevant after you saw the code.


I'm currently adding `__tenant__` claim to JWT token myself. If you see `ProfileService.cs`, at method `GetProfileDataAsync` I'm adding the tenant claim like this:

    claims.Add(new Claim("__tenant__", "initech"));

Default tenant is determined at `Startup.cs` `.WithStaticStrategy("finbuckle");`

While debugging `ApplicationDbContext` for `tenantInfo` it is always the default `finbuckle` tenant even why in the claim it is present `"__tenant__": "initech"` and `WithClaimStrategy` comes before `WithStaticStrategy`.

And the generated token is:

    eyJhbGciOiJSUzI1NiIsImtpZCI6IndsSlVsLW1OMTlkdlpxNnJJWGVvYkEiLCJ0eXAiOiJhdCtqd3QifQ.eyJuYmYiOjE2MDQ5MTgyODIsImV4cCI6MTYwNDkyMTg4MiwiaXNzIjoiaHR0cHM6Ly9sb2NhbGhvc3Q6NDQzNTAiLCJhdWQiOiJxdWlja2FwcF9hcGkiLCJjbGllbnRfaWQiOiJzd2FnZ2VydWkiLCJzdWIiOiI3ZGVhOTYzOC1iMzhmLTQyNGYtOWUyZS1kOWY2MWQwZjk5YmMiLCJhdXRoX3RpbWUiOjE2MDQ5MTgyODIsImlkcCI6ImxvY2FsIiwicm9sZSI6ImFkbWluaXN0cmF0b3IiLCJwZXJtaXNzaW9uIjpbInVzZXJzLnZpZXciLCJ1c2Vycy5tYW5hZ2UiLCJyb2xlcy52aWV3Iiwicm9sZXMubWFuYWdlIiwicm9sZXMuYXNzaWduIl0sIm5hbWUiOiJhZG1pbiIsImVtYWlsIjoiYWRtaW5AZWJlbm1vbm5leS5jb20iLCJwaG9uZV9udW1iZXIiOiIrMSAoMTIzKSAwMDAtMDAwMCIsImZ1bGxuYW1lIjoiSW5idWlsdCBBZG1pbmlzdHJhdG9yIiwiX190ZW5hbnRfXyI6ImluaXRlY2giLCJzY29wZSI6WyJxdWlja2FwcF9hcGkiXSwiYW1yIjpbInB3ZCJdfQ.SXXL08g1Dx7BUX59zLvnt_2QyNPabZMnigGzYaeghXjZOFGkC0jnSpRrD49IO3ub0tNcdZ9yUBjNyANaFe4UFfz_PeyKLCnOjJav6LQojk5n3wV45xJJwzuY1Xr9_0J9x8aPtcsEzZQPLe3hbAbJkazmk9KsMiIoLe83HHwprWNBQFtLAUpZh-9tbwVXuJF4EgSKpQtE18UNUZejBc5PFB4atuDa0ngbmSJqBkdTcg60w__QtpxIF6SEYx_3KrAl5BeWVHFiKS9Q5CvzO1K8I_8H8dqxMQoMuYFj9PmFIaD7C3gDDj_g7TBxwQDfi-3zxEf2GtNYjWF-9ntu9U6ocQ


In plain text looks like this (notice `__tenant__` claim is there):

    {
      "alg": "RS256",
      "kid": "wlJUl-mN19dvZq6rIXeobA",
      "typ": "at+jwt"
    }.{
      "nbf": 1604918282,
      "exp": 1604921882,
      "iss": "https://localhost:44350",
      "aud": "quickapp_api",
      "client_id": "swaggerui",
      "sub": "7dea9638-b38f-424f-9e2e-d9f61d0f99bc",
      "auth_time": 1604918282,
      "idp": "local",
      "role": "administrator",
      "permission": [
        "users.view",
        "users.manage",
        "roles.view",
        "roles.manage",
        "roles.assign"
      ],
      "name": "admin",
      "email": "admin@ebenmonney.com",
      "phone_number": "+1 (123) 000-0000",
      "fullname": "Inbuilt Administrator",
      "__tenant__": "initech",
      "scope": [
        "quickapp_api"
      ],
      "amr": [
        "pwd"
      ]
    }.[Signature]


I can't resolve tenant from claim. Am I missing something? 
Can I use `WithDelegateStrategy` and get the tenant with code from the claim?


**Update**:

I used `WithDelegateStrategy` to get tenant from `__tenat__` claim

	.WithDelegateStrategy(async context =>
	{
	    string tenantId = ((Microsoft.AspNetCore.Http.HttpContext)context).User?.FindFirst("__tenant__")?.Value;
	    return await Task.FromResult<string>(tenantId);
	})

But I need to put `app.UseMultiTenant();` below  `app.UseIdentityServer();` and `app.UseAuthorization();` in `Configure` method of `Startup` class otherwise I didn't get the claim as above.

However this change didn't make `WithClaimStrategy` to work.


**Update 2**:

Strategy isn't resolved when I use the endpoint to get token `connect/token`.  Debugger doesn't stop at breakpoint in `services.AddMultiTenant<ExtendedTenantInfo>()...` of `Startup` when I want to get a token.

I have modified configuration to look like this:

	services.AddMultiTenant<ExtendedTenantInfo>()
	.WithConfigurationStore()
	//.WithEFCoreStore<MultiTenantStoreDbContext, ExtendedTenantInfo>()                    
	.WithDelegateStrategy(async context =>
	{
	    string tenantId = ((Microsoft.AspNetCore.Http.HttpContext)context).User?.FindFirst("__tenant__")?.Value;
	    return await Task.FromResult<string>(tenantId);
	})
	.WithDelegateStrategy(async context =>
	{
	    ((Microsoft.AspNetCore.Http.HttpContext)context).Request.Headers.TryGetValue("tenant", out Microsoft.Extensions.Primitives.StringValues tenantId);  // get tenant from header "tenant"
	    return await System.Threading.Tasks.Task.FromResult(tenantId.ToString()); // ignore await warning or use await Task.FromResult(...)
	});

because first time that no tenant is available in claims I want to supply it from header request. 

What does prevent the strategy to get called in this case?

p.s. I have updated the repo with the changes I did.

**Comment by AndrewTriesToCode** [comment]
If I understand correctly you have `UseMultiTenant` after `UseIdentityServer`? I would recommend you try this order:

- UseAuthentication
- UseMultiTenant (it should be able to access `User` after authentication)
- UseAuthorization
- Use IdentityServer

Your delegate strategy is very similar to how the ClaimStrategy works-- yours is agnostic to the authentication scheme so works regardless of cookies or JWT. For technical reasons I had to change ClaimStrategy and the short version is I think in most apps it will only check for cookie claims because CookieAuthentication is the default authenticate scheme instead of OpenIdConnectAuthentication in the Microsoft templates. I will need to address this soon, but your solution is good.

I will say that `WithPerTenantAuthentication` will cause problems if you try to use it (that's why I had to change how ClaimStrategy works), but you probably don't  need it for API endpoint authentication.

**Comment by endrihacker** [comment]
Yes at the last try I did I left it like that. However in my template I used `UseAuthentication` isn't there at all. Does it get called by another method internally?

I tried the order that you suggested and also added `UseAuthentication` but notheing happend. I think that `UseAuthentication` doesn't still have the claims of the user and I can't determine a tenant from that.

My order that works look like below:

- UseAuthorization
- UseMultiTenant
- UseIdentityServer

I didn't use `UseAuthentication` at all.
I tried changing the order only in `Configure` method. Does it need to be changed even in `ConfigureServices` to make your sugestion work?

In default template `UseAuthorization` is after `UseIdentityServer`. Does it really matter? Does it bring any security issue?

Again thanks for your help :)

**Comment by AndrewTriesToCode** [comment]
hi again @endrihacker 

I apologize for how long this is taking! Can we back up and just try to reduce the problem?

Currently in the IdentityServer app the /Token endpoint is not getting the tenant from the JWT with the claim strategy right?

With respect to `UseAuthorization`, I believe thats after `UseIdentityServer` for your own endpoints that identity server ignores. For a token request it won't even hit that line.

I'm looking into how IdentityServer uses the authentication internally. We can make this work I just need to understand what it's doing internally.

**Comment by AndrewTriesToCode** [comment]
Ok, now that I've refreshed myself I realized that in OpenId Connect when a request is made to the token endpoint the authentication code (in the form of a JWT) is send as form encoded body data.

This means that authentication never looks at it which is why we aren't seeing the claim as part of the user. I checked the IdentityServer source and it doesn't do JWT authentication itself for this endpoint, but it does reads the form data an interprets it as a JWT. Simply put the /connect/token endpoint IdentityServer users doesn't actually use the JWT you pass it for authentication.

You're making me face some hard questions with Identity Server 4 I've been trying to put off!

After a token is obtained from IdentityServer and you want to authenticate it on your own API endpoints the stuff we've been talking about comes into play.

**Comment by endrihacker** [comment]
Hello @AndrewTriesToCode 

I wanted to say that I made those two delegate strategies work just by using this order

- UseAuthorization
- UseMultiTenant
- UseIdentityServer

I noticed that If I put `UseMultiTenant` after `UseAuthorization` and before `UseIdentityServer` it works (by works I mean I can determine `tenantId` in both delegate methods). That's why I can't use `UseAuthorization` after `UseIdentityServer`.

My logic is like this:
user need to login to access protected API endpoints and at the login phase I get the tenant from what users says (put it on header). Every tenant has its own database. After successful authentication I add to claims the tenant. And in each other request I get the tenant from the claim in access token.

I'll keep it like this for now and see later if I find a better approach. This seems to work for now.
